### ISSUE TITLE
[LLVM 18+][OpenMP] OpenMP target offloading will cause missed CUPTI callbacks

### ISSUE THREAD

**Issue by Thyre** [llvm/llvm-project#85770]
[LLVM 18+][OpenMP] OpenMP target offloading will cause missed CUPTI callbacks
===

## Introduction to the issue

Analyzing performance of applications is important to get insigt into further optimization potential. There exist several different tools to look into this. For NVIDIA GPUs, both Nsight Compute and Nsight Systems exist. With both these tools, users can quickly get insight into their application and get some optimization recommendations. 

In https://github.com/llvm/llvm-project/pull/74397, it was reported that Nsight Compute shows issues with several different LLVM versions. Here, it was noted that LLVM calls CUDA functions from static initializers, which is not allowed as specified in the documentation:

> The CUDA interfaces use global state that is initialized during host program initiation and destroyed during host program termination. The CUDA runtime and driver cannot detect if this state is invalid, so using any of these interfaces (implicitly or explicitly) during program initiation or termination after main) will result in undefined behavior.

In addition, other tools like Score-P suffer issues due to the early initialization of `libomptarget.so` which also initializes the OpenMP Tools interface. More information can be found in https://github.com/llvm/llvm-project/pull/74397. Basically, the OpenMP Tools Interface was initialized during `_dl_init`. If we tried to initialize other adapters, like CUDA or rocProfiler/rocTracer/rocm_smi, we would end up with a segmentation fault. This issue has been resolved from our side, where we now try to initialize adapters at a later point in `_start`.

## Changes to the initialization since LLVM 18

In https://github.com/llvm/llvm-project/pull/74397, changes to the device initialization were made. As noted after the PR was merged, all available devices are getting initialized eagerly through `PluginAdaptorTy::initDevices` now. While this works fine for the OpenMP Tools Interface, this has certainly an impact on tools using CUPTI to analyze user applications. 

Lets look at a simple example code:

```c
#include <stdio.h>
#include <omp.h>

int main( void )
{
    #pragma omp target teams num_teams(2)
	{
	    printf("omp_is_initial_device() = %d | omp_get_team_num() = %d\n", omp_is_initial_device(), omp_get_team_num());
	}
}
```

Running the user application normally yields these results

```console
$ clang -fopenmp --offload-arch=native ./reproducer.c
$ ./a.out
omp_is_initial_device() = 0 | omp_get_team_num() = 1
omp_is_initial_device() = 0 | omp_get_team_num() = 0
```

Trying to run the same application with Nsight Compute fails as expected. We see the following error message:

```console
$ ncu ./a.out
==ERROR== Cuda is initialized before the tool, e.g. by calling a Cuda API from a static initializer.
==ERROR== Initializing Cuda during program initialization results in undefined behavior.
==ERROR== See https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#initialization
==ERROR== The application returned an error code (1).
==WARNING== No kernels were profiled.
```

These results are all known and reported in the mentioned issues above. Now for the new part.

### CUDA calls happening before performance tools can initialize

With the changes in the mentioned PR, LLVM will initialize devices extremely early during program execution. This includes a call to `cuStreamCreate`. See this backtrace for more infomation:

```gdb
#0  0x00007ffff3f95c60 in cuStreamCreate(CUstream_st**, unsigned int) ()
   from /home/jreuter/Projects/Compilers/llvm-project/_build/_install/lib/x86_64-unknown-linux-gnu/libomptarget.rtl.cuda.so
#1  0x00007ffff3f81ff7 in llvm::omp::target::plugin::CUDAStreamRef::create(llvm::omp::target::plugin::GenericDeviceTy&) ()
   from /home/jreuter/Projects/Compilers/llvm-project/_build/_install/lib/x86_64-unknown-linux-gnu/libomptarget.rtl.cuda.so
#2  0x00007ffff3f81be5 in llvm::omp::target::plugin::GenericDeviceResourceManagerTy<llvm::omp::target::plugin::CUDAStreamRef>::resizeResourcePoolImpl(unsigned int, unsigned int) () from /home/jreuter/Projects/Compilers/llvm-project/_build/_install/lib/x86_64-unknown-linux-gnu/libomptarget.rtl.cuda.so
#3  0x00007ffff3f81889 in llvm::omp::target::plugin::GenericDeviceResourceManagerTy<llvm::omp::target::plugin::CUDAStreamRef>::resizeResourcePool(unsigned int) () from /home/jreuter/Projects/Compilers/llvm-project/_build/_install/lib/x86_64-unknown-linux-gnu/libomptarget.rtl.cuda.so
#4  0x00007ffff3f7db96 in llvm::omp::target::plugin::CUDADeviceTy::initImpl(llvm::omp::target::plugin::GenericPluginTy&) ()
   from /home/jreuter/Projects/Compilers/llvm-project/_build/_install/lib/x86_64-unknown-linux-gnu/libomptarget.rtl.cuda.so
#5  0x00007ffff3f9a3dd in llvm::omp::target::plugin::GenericDeviceTy::init(llvm::omp::target::plugin::GenericPluginTy&) ()
   from /home/jreuter/Projects/Compilers/llvm-project/_build/_install/lib/x86_64-unknown-linux-gnu/libomptarget.rtl.cuda.so
#6  0x00007ffff3fa04b0 in llvm::omp::target::plugin::GenericPluginTy::initDevice(int) ()
   from /home/jreuter/Projects/Compilers/llvm-project/_build/_install/lib/x86_64-unknown-linux-gnu/libomptarget.rtl.cuda.so
#7  0x00007ffff3fa0bab in __tgt_rtl_init_device ()
   from /home/jreuter/Projects/Compilers/llvm-project/_build/_install/lib/x86_64-unknown-linux-gnu/libomptarget.rtl.cuda.so
#8  0x00007ffff5edfbfd in DeviceTy::init() ()
   from /home/jreuter/Projects/Compilers/llvm-project/_build/_install/lib/x86_64-unknown-linux-gnu/libomptarget.so.19.0git
#9  0x00007ffff5efb8ed in PluginAdaptorTy::initDevices(PluginManager&) ()
   from /home/jreuter/Projects/Compilers/llvm-project/_build/_install/lib/x86_64-unknown-linux-gnu/libomptarget.so.19.0git
#10 0x00007ffff5efc169 in PluginManager::registerLib(__tgt_bin_desc*) ()
   from /home/jreuter/Projects/Compilers/llvm-project/_build/_install/lib/x86_64-unknown-linux-gnu/libomptarget.so.19.0git
#11 0x000055555555618d in omp_offloading.descriptor_reg ()
#12 0x00007ffff3c29ebb in call_init (env=<optimized out>, argv=0x7fffffffa788, argc=1) at ../csu/libc-start.c:145
#13 __libc_start_main_impl (main=0x555555556290 <main>, argc=1, argv=0x7fffffffa788, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, 
    stack_end=0x7fffffffa778) at ../csu/libc-start.c:379
#14 0x00005555555561c5 in _start ()
```

The initial caller is `omp_offloading.descriptor_reg`. This function is created [here](https://github.com/llvm/llvm-project/blob/d9c31ee9568277e4303715736b40925e41503596/llvm/lib/Frontend/Offloading/OffloadWrapper.cpp#L194) with a priority of `1`. The reason for this is given as well:
```c
  // Add this function to constructors.
  // Set priority to 1 so that __tgt_register_lib is executed AFTER
  // __tgt_register_requires (we want to know what requirements have been
  // asked for before we load a libomptarget plugin so that by the time the
  // plugin is loaded it can report how many devices there are which can
  // satisfy these requirements).
```

As the initialization is happening this early, a performance infrastructure tool has no chance to catch this. With Score-P, one can observe the following result:

```
$ scorep --cuda --thread=omp:ompt clang -fopenmp --offload-arch=native reproducer.c
$ ./a.out
[Score-P] src/adapters/cuda/scorep_cupti4_activity.c:572: Warning: [CUPTI Activity] Context not found! 
omp_is_initial_device() = 0 | omp_get_team_num() = 1
omp_is_initial_device() = 0 | omp_get_team_num() = 0
[Score-P] src/adapters/cuda/scorep_cupti4_activity.c:131: Warning: [CUPTI Activity] Context not found! Cannot flush context ...
[Score-P] src/measurement/scorep_definition_cube4.c:771: Warning: Given metric name "CUDA Memory" was changed to "CUDA_Memory" for CubePL processing, i.e., .spec file and CubeGUI derived metrics processing. Given name still used for display. Note, profiling only.
```

Score-P requries information about created streams to correctly associate and evaluate buffers allocated for the CUPTI adapter. As `cuStreamCreate` is happening before Score-P has a chance to initialize CUPTI, we will miss this function call. This yields to OpenMP target events missing from our measurements entirely. 

![image](https://github.com/llvm/llvm-project/assets/14841361/a1e91633-272f-48bb-a194-f7a696130fcc)

We can compare this to LLVM 17.0.6, where everything is working perfectly fine.

![image](https://github.com/llvm/llvm-project/assets/14841361/c8a1c124-871f-4b11-af69-b62e403bdceb)

From our side, there's nothing we can do about that. Initializing any earlier would mean that we're in `_dl_init` again, where initializing CUPTI will simply cause a segmentation fault. There may be a way around the missed stream creation by trying to do our handling when detecting we missed it, but disregards other CUDA function calls we missed as well and is not guaranteed to work.

### Appendix: Build Score-P to reproduce issue

Score-P v8.4 does not contain the fix to allow using CUDA and our OMPT adapter simultaneously. One can do the following steps to build Score-P master and reproduce the issue:

```console
$ wget https://perftools.pages.jsc.fz-juelich.de/cicd/scorep/branches/master/latest.tar.gz
$ tar -xf latest.tar.gz --strip-components=1
$ mkdir _build && cd _build
$ NVCC="nvcc -allow-unsupported-compiler" ../configure --enable-debug --enable-shared --with-nocross-compiler-suite=clang --without-mpi --without-shmem --prefix=$(pwd)/_install
$ make && make install
$ export PATH=$(pwd)/_install/bin:${PATH}
```

**Comment by jhuber6** [comment]
Is this still an issue?

**Comment by Thyre** [comment]
I'm sorry for the very late answer.

We have started working on an entirely new CUDA adapter. There, we are able to record all activities which happen _after_ we are able to attach our tool, e.g. on the first adapter event or our static constructor. With this, we are able to do measurements at all and get all the user kernels.

What is still missing are host-side API calls, which happen before we are able to attach Score-P. While this is fine for the most part, I've noticed today that we will lose the information about the device (see [CUpti_ActivityDevice5](https://docs.nvidia.com/cupti/api/structCUpti__ActivityDevice5.html?highlight=CUpti_ActivityDevice5#_CPPv421CUpti_ActivityDevice5)) because of that.

If a tool wants to record information via `cuptiActivityEnable( CUPTI_ACTIVITY_KIND_DEVICE )`, this call has to be done before `cuInit( 0 )`. If not, then no activity record of the type `CUPTI_ACTIVITY_KIND_DEVICE` will be dispatched. Because of the very early initialization of the CUDA driver by LLVM/Clang, we encounter this case for all OpenMP programs using offloading. I verified this with LLVM 19.1.0-rc3. 

So basically, the issue will not be as severe once we finish our new CUDA adapter (which will certainly take a while), but its still inconvenient.

**Comment by jhuber6** [comment]
If this is an ordering thing, presumably you could document the priority the constructors for your hook run at so we can ensure that the OpenMP ones are run later. We currently run them at `101` which is the earliest approved by the C library, but could bump it back if needed. Generally we just want to run this before normal C++ constructors, because those could contain OpenMP calls themselves. (Those are 65535 priorty).

**Comment by Thyre** [comment]
This indeed just looks like an ordering problem. For testing, I moved up the priority of `omp_offloading.descriptor_reg` just a tiny bit:

```diff
diff --git a/llvm/lib/Frontend/Offloading/OffloadWrapper.cpp b/llvm/lib/Frontend/Offloading/OffloadWrapper.cpp
index 8b6f9ea1f4cc..0dc6a7fa5d85 100644
--- a/llvm/lib/Frontend/Offloading/OffloadWrapper.cpp
+++ b/llvm/lib/Frontend/Offloading/OffloadWrapper.cpp
@@ -242,7 +242,7 @@ void createRegisterFunction(Module &M, GlobalVariable *BinDesc,
   Builder.CreateRetVoid();
 
   // Add this function to constructors.
-  appendToGlobalCtors(M, Func, /*Priority=*/101);
+  appendToGlobalCtors(M, Func, /*Priority=*/102);
 }
 
 // struct fatbin_wrapper {
```

I then rebuilt LLVM 19.1.0-rc3 and changed the Score-P constructor to use a priority of `101`. With this, I was able to get all the events.

Prior to my changes:
![Screenshot from 2024-09-06 18-46-46](https://github.com/user-attachments/assets/48b17c6c-42aa-414c-941c-b76aebc7880a)

With the changes:
![Screenshot from 2024-09-06 18-45-58](https://github.com/user-attachments/assets/c36ebb74-a91b-4b73-8b83-a47571d2f91b)

So I'd guess having a (very small) window where tools can initialize themselves would be sufficient for us. Keeping a priority close to `101` shouldn't cause issues with most normal constructors.

**Comment by Thyre** [comment]
The merge request for Score-P is stuck in review for some time due to other things having priority, but it will very likely include the following change:

```
- Initialization of Score-P via static constructors will now use
  a priority of 101 for `__attribute__((constructor))`, if supported
  by the compiler. This is to ensure that Score-P is initialized
  before other constructors that might dispatch events.
```

We will improve the documentation for this (probably in a separate file for developers and so on, but this will at least be in our `ChangeLog` for now).
